lib/core/src/Cardano/Wallet/DB/Store/Transactions/Model.hs

```diff
@@ -0,0 +1,280 @@
+import Cardano.Wallet.Primitive.Types.TokenQuantity
+    ( TokenQuantity )
+import Control.Arrow
+    ( (&&&) )
+import Data.Delta
+    ( Delta (..) )
+import Data.Foldable
+    ( fold )
+import Data.Generics.Internal.VL
+    ( view, (^.) )
+import Data.List
+    ( sortOn )
+import Data.Map.Strict
+    ( Map )
+import Data.Word
+    ( Word32 )
+import Fmt
+    ( Buildable (build) )
+import GHC.Generics
+    ( Generic )
+
+import qualified Cardano.Wallet.Primitive.Types.Coin as W
+import qualified Cardano.Wallet.Primitive.Types.TokenBundle as TokenBundle
+import qualified Cardano.Wallet.Primitive.Types.Tx as W
+import qualified Data.Map.Strict as Map
+
+-- | A context that carries a TxOut together with its tokens
+-- (this will be needed in the future for the DB Layer).
+data WithTxOut a =
+    WithTxOut
+    { withTxOut_value :: a, withTxOut_context :: Maybe (TxOut, [TxOutToken]) }
+    deriving ( Show, Eq, Functor )
+
+-- | A kind to index the 2 flavours of a 'TxRelationF', with or without 'TxOuts'
+data TxRelationTxOutCtxK
+    = NoTxOutCtx
+    | WithTxOutCtx
+
+-- | Define the TxOut context type
+type family TxRelationTxOutCtx f a where
+    TxRelationTxOutCtx 'NoTxOutCtx a = a
+    TxRelationTxOutCtx 'WithTxOutCtx a = WithTxOut a
+
+{- | A low level definition of a transactions covering all transaction content
+ by collecting all related-to-index database rows.
+
+ Normalization is performed anyway after the first realtion level.
+
+ All values used here are records in the database: all foreign keys are used
+ to group data correctly but they are not removed from the data.
+-}
+data TxRelationF (f :: TxRelationTxOutCtxK) =
+    TxRelationF
+    { txRelation_ins :: [TxRelationTxOutCtx f TxIn]
+    , txRelation_colls :: [TxRelationTxOutCtx f TxCollateral]
+    , txRelation_outs :: [(TxOut, [TxOutToken])]
+    , txRelation_collouts :: Maybe (TxCollateralOut, [TxCollateralOutToken])
+    , txRelation_withdraws :: [TxWithdrawal]
+    }
+    deriving ( Generic )
+
+deriving instance ( Eq (TxRelationTxOutCtx f TxIn)
+                  , Eq (TxRelationTxOutCtx f TxCollateral))
+    => Eq (TxRelationF f)
+
+deriving instance ( Show (TxRelationTxOutCtx f TxIn)
+                  , Show (TxRelationTxOutCtx f TxCollateral))
+    => Show (TxRelationF f)
+
+-- | Transactions history is 'TxRelationF's indexed by 'TxId'
+newtype TxHistoryF f =
+    TxHistoryF { txHistory_relations :: Map TxId (TxRelationF f) }
+    deriving ( Generic )
+
+deriving instance ( Eq (TxRelationTxOutCtx f TxIn)
+                  , Eq (TxRelationTxOutCtx f TxCollateral))
+    => Eq (TxHistoryF f)
+
+deriving instance ( Show (TxRelationTxOutCtx f TxIn)
+                  , Show (TxRelationTxOutCtx f TxCollateral))
+    => Show (TxHistoryF f)
+
+instance Monoid (TxHistoryF f) where
+    mempty = TxHistoryF mempty
+
+instance Semigroup (TxHistoryF f) where
+    TxHistoryF h1 <> TxHistoryF h2 =
+        TxHistoryF $ Map.unionWith (error "clash") h1 h2
```

Comment:
```suggestion
        TxHistoryF $ Map.unionWith (\_ r -> r) h1 h2
```

The `(<>)` operation is intended to be total — we can't afford to call `error` here. This definition should work, because if two transactions have the same `TxId`, they should be identical. (If they are not identical, that's an error — but we blame some other part of the code, not this part.)

(Using a right-biased union should be correct when combined with the current definition of `apply (Expand txs)` and a definition of `updateS` using `repsertMany`.)